{
  "comments": [
    {
      "unresolved": true,
      "key": {
        "uuid": "16df91f3_cbbdd84d",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1000153
      },
      "writtenOn": "2024-10-14T17:00:16Z",
      "side": 1,
      "message": "This patch needed rebased, and had some lifetime issues as written.  I\u0027ve rewritten some portions to set up shared_from_this so we get proper lifetimes for subscriptions on delete that might not have all their connections closed.\n\nI haven\u0027t tested it, so PTAL and see if it passes your tests.",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "8b1d55f5_9d1a683f",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1001604
      },
      "writtenOn": "2024-10-14T22:42:12Z",
      "side": 1,
      "message": "I found out there is a bug in https://gerrit.openbmc.org/c/openbmc/bmcweb/+/65720 and I submitted a fix https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75296.\n\nI\u0027ll rebase it when ready",
      "parentUuid": "16df91f3_cbbdd84d",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "41734a6b_df9558a8",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1001604
      },
      "writtenOn": "2024-10-15T23:19:37Z",
      "side": 1,
      "message": "With this, I\u0027m seeing the bad_weak_ptr like\n```\nOct 15 22:48:40 p10bmc bmcwebd[6227]: terminate called after throwing an instance of \u0027std::bad_weak_ptr\u0027\nOct 15 22:48:40 p10bmc bmcwebd[6227]:   what():  bad_weak_ptr\nOct 15 22:48:41 p10bmc systemd[1]: bmcweb.service: Main process exited, code\u003ddumped, status\u003d6/ABRT\nOct 15 22:48:41 p10bmc systemd[1]: bmcweb.service: Failed with result \u0027core-dump\u0027.\nOct 15 23:03:04 p10bmc systemd[1]: Started Start bmcwebd server.\n```\n\nI think this seems related to the class `Subscription` and there seems a mismatch when it is shared.\n\nI\u0027ll keep looking at it.\n\n```\nclass Subscription : std::enable_shared_from_this\u003cSubscription\u003e\n\n...\n\n            client-\u003esendDataWithCallback(\n....\n                std::bind_front(\u0026Subscription::resHandler, this,\n                                shared_from_this()));\n                                \n                                \n```",
      "parentUuid": "8b1d55f5_9d1a683f",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "2070043b_38e87794",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1000020
      },
      "writtenOn": "2024-10-18T17:44:20Z",
      "side": 1,
      "message": "https://gerrit.openbmc.org/c/openbmc/bmcweb/+/75296 merged. So I think we are good here?",
      "parentUuid": "41734a6b_df9558a8",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "03849138_f6164ec8",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1001604
      },
      "writtenOn": "2024-10-18T18:16:37Z",
      "side": 1,
      "message": "Even after rebase, this still does not work.\n\n1. When using `shared_from_this()` (e.g. patchset 11), it caused `bad_weak_ptr` exception like\n```\nOct 15 22:48:40 p10bmc bmcwebd[6227]: terminate called after throwing an instance of \u0027std::bad_weak_ptr\u0027\nOct 15 22:48:40 p10bmc bmcwebd[6227]:   what():  bad_weak_ptr\n```\n\nI think in the process of this code path, Subscription shared_ptr may have been released.\n\n2. Even with the change as the current patchset 14, it still fails to cause SEQV like\n```\nOct 18 17:01:08 p10bmc systemd[1]: bmcweb.service: Main process exited, code\u003ddumped, status\u003d11/SEGV\nOct 18 17:01:08 p10bmc systemd[1]: bmcweb.service: Failed with result \u0027core-dump\u0027.\n```\n\n----\nI\u0027ll keep looking",
      "parentUuid": "41734a6b_df9558a8",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "19d171d5_0d2b3a28",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1000153
      },
      "writtenOn": "2024-10-18T18:36:17Z",
      "side": 1,
      "message": "where does that exception get thrown from?",
      "parentUuid": "03849138_f6164ec8",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "04526c6c_50cda3c9",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1001604
      },
      "writtenOn": "2024-10-18T19:25:25Z",
      "side": 1,
      "message": "With patcheset 12 (and with rebase on top of the current master) , the exception comes from `shared_from_this()` here.\n\n```\n            client-\u003esendDataWithCallback(\n....\n                std::bind_front(\u0026Subscription::resHandler, this,\n                                shared_from_this()));\n```\nTo prove it, add the line like\n\n```\n@@ -318,6 +318,9 @@ class Subscription : std::enable_shared_from_this\u003cSubscription\u003e\n\n         if (client)\n         {\n+            BMCWEB_LOG_ERROR(\"Before shared_from_this\");\n+            auto self \u003d shared_from_this();\n+            BMCWEB_LOG_ERROR(\"After shared_from_this\");\n             client-\u003esendDataWithCallback(\n                 std::move(msg), userSub.destinationUrl,\n                 static_cast\u003censuressl::VerifyCertificate\u003e(\n```\n\nThen, an exception throw is observed here when a new event is occurred.\n\n```\nOct 18 19:17:02 p10bmc bmcwebd[286]: [DEBUG http_connection.hpp:302] Setting completion handler\nOct 18 19:17:02 p10bmc bmcwebd[286]: [DEBUG http_response.hpp:258] 0x2688b90 setting completion handler\nOct 18 19:17:02 p10bmc bmcwebd[286]: [DEBUG routing.hpp:653] Matched rule \u0027/redfish/v1/EventService/Actions/EventService.SubmitTestEvent/\u0027 POST / 32\nOct 18 19:17:02 p10bmc bmcwebd[286]: [DEBUG dbus_privileges.hpp:50] userName \u003d root userRole \u003d priv-admin\nOct 18 19:17:02 p10bmc bmcwebd[286]: [DEBUG query.hpp:123] setup redfish route\nOct 18 19:17:02 p10bmc bmcwebd[286]: [ERROR event_service_manager.hpp:321] Before shared_from_this\nOct 18 19:17:02 p10bmc bmcwebd[286]: terminate called after throwing an instance of \u0027std::bad_weak_ptr\u0027\nOct 18 19:17:02 p10bmc bmcwebd[286]:   what():  bad_weak_ptr\n```",
      "parentUuid": "19d171d5_0d2b3a28",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "d5bba15a_a515a3f6",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1001604
      },
      "writtenOn": "2024-10-20T18:20:17Z",
      "side": 1,
      "message": "It turns out the cause is due to Subscription with `private` enable_shared_from_this inheritance.  (Some more reference is https://stackoverflow.com/questions/39937112/stdenable-shared-from-this-public-vs-private).\n\nThe fix is to do like\n```\nclass Subscription : std::enable_shared_from_this\u003cSubscription\u003e\n--\u003e\nclass Subscription : public std::enable_shared_from_this\u003cSubscription\u003e\n```\n\nNow it works.",
      "parentUuid": "04526c6c_50cda3c9",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "ef28c652_9aad7995",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1001604
      },
      "writtenOn": "2024-10-20T21:27:56Z",
      "side": 1,
      "message": "I\u0027d like to do more testing as I observed some missed deletion of subscription after TerminateAfterRetries\n```",
      "parentUuid": "d5bba15a_a515a3f6",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "e89c8a21_46504894",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1001604
      },
      "writtenOn": "2024-10-20T21:53:42Z",
      "side": 1,
      "message": "There was a missing `deleter` setup at Subscription PATCH (for push style).\nIt should now be fixed and working.",
      "parentUuid": "ef28c652_9aad7995",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    },
    {
      "unresolved": true,
      "key": {
        "uuid": "b2894b95_4027bd8f",
        "filename": "/PATCHSET_LEVEL",
        "patchSetId": 7
      },
      "lineNbr": 0,
      "author": {
        "id": 1001604
      },
      "writtenOn": "2024-10-20T23:10:17Z",
      "side": 1,
      "message": "I meant - POST (subscription creation)",
      "parentUuid": "e89c8a21_46504894",
      "revId": "3da1f6c3309be413e106afbc5ff1a9f7a14bbb7e",
      "serverId": "adbd0a64-1f21-4d83-b585-671fe73cb6e4"
    }
  ]
}